Userland/Libraries/LibSoftGPU/Image.cpp

```diff
@@ -100,44 +99,114 @@ void Image::write_texels(u32 layer, u32 level, Vector3<i32> const& output_offset
 }
 
-void Image::copy_texels(GPU::Image const& source, u32 source_layer, u32 source_level, Vector3<u32> const& source_offset, Vector3<u32> const& size, u32 destination_layer, u32 destination_level, Vector3<u32> const& destination_offset)
+void Image::copy_texels(GPU::Image const& source, u32 source_level, Vector3<u32> const& source_offset, Vector3<u32> const& size, u32 destination_level, Vector3<u32> const& destination_offset)
 {
     VERIFY(source.has_same_ownership_token(*this));
 
     auto const& src_image = static_cast<Image const&>(source);
 
-    VERIFY(source_layer < src_image.num_layers());
-    VERIFY(source_level < src_image.num_levels());
-    VERIFY(source_offset.x() + size.x() <= src_image.level_width(source_level));
-    VERIFY(source_offset.y() + size.y() <= src_image.level_height(source_level));
-    VERIFY(source_offset.z() + size.z() <= src_image.level_depth(source_level));
-    VERIFY(destination_layer < num_layers());
-    VERIFY(destination_level < num_levels());
-    VERIFY(destination_offset.x() + size.x() <= level_width(destination_level));
-    VERIFY(destination_offset.y() + size.y() <= level_height(destination_level));
-    VERIFY(destination_offset.z() + size.z() <= level_depth(destination_level));
+    VERIFY(source_level < src_image.number_of_levels());
+    VERIFY(source_offset.x() + size.x() <= src_image.width_at_level(source_level));
+    VERIFY(source_offset.y() + size.y() <= src_image.height_at_level(source_level));
+    VERIFY(source_offset.z() + size.z() <= src_image.depth_at_level(source_level));
+    VERIFY(destination_level < number_of_levels());
+    VERIFY(destination_offset.x() + size.x() <= width_at_level(destination_level));
+    VERIFY(destination_offset.y() + size.y() <= height_at_level(destination_level));
+    VERIFY(destination_offset.z() + size.z() <= depth_at_level(destination_level));
 
     for (u32 z = 0; z < size.z(); ++z) {
         for (u32 y = 0; y < size.y(); ++y) {
             for (u32 x = 0; x < size.x(); ++x) {
-                auto color = src_image.texel(source_layer, source_level, source_offset.x() + x, source_offset.y() + y, source_offset.z() + z);
-                set_texel(destination_layer, destination_level, destination_offset.x() + x, destination_offset.y() + y, destination_offset.z() + z, color);
+                auto color = src_image.texel(source_level, source_offset.x() + x, source_offset.y() + y, source_offset.z() + z);
+                set_texel(destination_level, destination_offset.x() + x, destination_offset.y() + y, destination_offset.z() + z, color);
             }
         }
     }
 }
 
+static GPU::ImageDataLayout image_data_layout_for_bitmap(Gfx::Bitmap& bitmap)
+{
+    return GPU::ImageDataLayout {
+        .pixel_type = {
+            .format = GPU::PixelFormat::BGRA,
+            .bits = GPU::PixelComponentBits::B8_8_8_8,
+            .data_type = GPU::PixelDataType::UnsignedInt,
```

Comment:
Reminder to self: check why this is different from the backing image in `Device`.